File: extract_genes.pl
#!/usr/bin/env perl
#By Thomas J Creedy, thomas@tjcreedy.co.uk 

use warnings;
use strict;
use Data::Dumper;
use Getopt::Long;
use Bio::SeqIO;
use File::Path qw(make_path);
use File::Basename;
use Array::Utils qw(intersect unique);
use List::MoreUtils qw(uniq);
use LWP::Simple;

my @gbpaths;
my $outdir;

my $minregion = 1;
my @reqregion;
my @regiontypes = ('CDS');
my $keepframe;
my $organism;
my $writeunknowns;

my $dontchecknames;
my $showgenes;
my $help;

my $script = basename($0, ());

my $genenames = get 'https://raw.githubusercontent.com/tjcreedy/biotools/master/gene_name_variants.txt';

my ($known_genes, $genesort, $generegion) = sort_genes($genenames);

print Dumper $genesort;
exit;

########################################################
# USAGE
#
my $usage =<<USAGE;

Description:
	This script finds and extracts sequences corresponding to CDS, rRNA and/or tRNA annotations from a set of sequences from one or more genbank-format flat files. Specify which regions to extract using the -regiontypes argument (default is -regiontypes CDS rRNA)
	
	The region name is identified based on the /gene, /label or /product tag in the genbank-format flat file. Two methods are used to ensure that naming variants are correctly identified as the same gene:
		1. the gene name is converted to uppercase (so that "atp8" is the same as "ATP8")
		2. the script removes any semicolons, underscores, hyphens or spaces, and any characters following these, from gene names ("ATP8-0" -> "ATP8", "ATP8; atp8" -> "ATP8", etc) - except if the first part of the name is "MT-".
	
	Additionally, common naming variants of mitochondrial regions are hardcoded into the script (e.g. CDS regions COX1 and COI will both be output to COX1). The variants that the script knows about can be reviewed using the -showgenes option.
	
	A separate fasta will be written for each region found, containing all of the sequences of that CDS from across all the different sequences in the genbank file(s). The LOCUS name of the sequence in the genbank-format flat file will be used as the sequence header. Alternatively, you can try to use use the organism field of the genbank file as the sequence header using --organism: note that this is very likely to give blank or non-unique headers, and if the script fails to parse the species name it will revert to the LOCUS field.
	
	Optionally, you can specify that sequences will only be output for entries where a minimum number of regions are met using -minregion. Similarly, you can optionally specify that certain regions must be present in a sequence for it to output any regions using the -reqregion option. For example, if you only wanted to retrieve sequences for regions from genomes that contain COX1 and COB, you would specify -reqregion COX1 COB. The names used must conform with the standard set of output names (the second column when using the -showgenes option). All available regions will still be output.
	
	In rare cases, an annotation is truncated by the end of a contig at the 5' end, which may leave an incomplete codon at the start of an output sequence and cause downstream issues. To detect this, the script can optionally remove these excess 1 or 2 bases using the -keepframe option. Note this only works if the /codon_start tag is present in the genbank-format features. 

Usage:

	perl $script -genbank <in.gb> [<in2.gb>] -out <out/> [-minregion <n>] [-reqregion <CDS1> <CDS2>] [-regiontypes CDS tRNA rRNA] [-keepframe]
	perl $script -showgenes

Arguments:

	       genbank:  The path to one or more genbank files from which CDS regions should be extracted
	           out:  The path to a directory in which to place a fasta file for each CDS
	     minregion:  A minimum number of regions that must be present in a sequence for it to output anything
	     reqregion:  A list of regions that must be present in a sequence for it to output anything
	   regiontypes:  One or more region types to extract, currently CDS rRNA and/or tRNA
	     showgenes:  Prints the hardcoded conversions for region naming variants
	      organism:  Prints the organism name, if retreivable, instead of the LOCUS
	 writeunknowns:  Writes out fastas for unidentifiable annotations anyway
	     keepframe:  Removes excess out-of-frame bases at the beginning of truncated annotations
	          help:  Prints out this helpful message

USAGE
#
######################################################


GetOptions("genbank=s{1,}"	=> \@gbpaths,
	   "out=s"		=> \$outdir,
	   "minregion=i"	=> \$minregion,
	   "reqregion=s{0,}"	=> \@reqregion,
	   "regiontypes=s{1,3}"	=> \@regiontypes,
	   "showgenes"		=> \$showgenes,
	   "organism"		=> \$organism,
	   "writeunknowns"	=> \$writeunknowns,
	   "keepframe"		=> \$keepframe,
	   "help"		=> \$help) or die "\nError getting options\n\n";

print $usage and exit if $help;

print "\t\'VARIANT\' => \'OUTPUT\'\n", Dumper $genesort and exit if $showgenes;

die "Error: please supply the path(s) to one or more genbank files and the path to an output directory\n" unless @gbpaths and $outdir;

# Check all values of reqregion make sense
die "Error: one or more region names given to -reqregion don't match any known gene names\n" unless scalar (intersect(@reqregion, @$known_genes)) == scalar @reqregion;

# Check all values of regiontypes make sense
my @alltypes = ('CDS', 'rRNA', 'tRNA');
die "Error: supply one or more region types to -regiontypes\n" unless scalar @regiontypes > 0;
die "Error: one or more region types given to -regiontypes don't match any known region types\n" unless scalar (unique(@regiontypes, @alltypes)) == 3;

my %regtypes = map {$_ => 1} @regiontypes;

# Make output directory
make_path($outdir);

# Set up fasta objects for outputs
my %faobjs;

# Set up hash for unrecognised gene names
my %unrec_genes;

# Work through genbank files
foreach my $gbp (@gbpaths){
	
	# Initialise genbank read object
	my $gb_in = Bio::SeqIO->new(-file => $gbp,
				    -format => "genbank",
				    -verbose => -1);
	
	# Work through sequences in object
	while(my $seq = $gb_in->next_seq){
		
		$seq->verbose(-1);
		
		# Extract sequence name and species name
		my $seqname = $seq->display_id;
		my $species;
		if($organism){
			$species = $seq->species->node_name;
			$species =~ tr/ /_/;
		}
		# Get all features with sufficient identification from object
		my @all_feats = grep {$_->has_tag('gene') or $_->has_tag('product') or $_->has_tag('label')} ($seq->get_SeqFeatures);
		
		# Set up containers
		my %found_genes;
		my $checked_features = 0;
		
		#printf "Seqname = $seqname, Feats = %d, max_checked_feats = %d\n", $#all_feats, $#all_feats * 2 + 1;
		
		# Work through features to extract data
		while($checked_features < ($#all_feats * 2 + 1) and scalar keys %found_genes < scalar @all_feats){
			
			# Assess current position in search
			my $round = $checked_features <= $#all_feats ? 0 : 1;
			my $i = $round ? $checked_features - $#all_feats - 1 : $checked_features;
			
			# Extract feature
			my $feat = $all_feats[$i];
			$feat->verbose(-1);
			
			my $feattype = $feat->primary_tag;
			
			# Get clean gene name and convert
			my $gene = get_clean_name($feat);
			$gene = ${$genesort}{$gene} if ${$genesort}{$gene};
			
			#print "Checked_features = $checked_features, Round = $round, i = $i, feattype = $feattype, gene = $gene ...";
			
			
			# If gene has not already been extracted, it is the first round of checks and this is a required type of feature or it is the second round of checks and this is a gene feature and the gene type matches those required
			
			if( not $found_genes{$gene} and 
					( $regtypes{$feattype} or (
						$feat->primary_tag eq 'gene' and 
						$round and 
						${$generegion}{$gene} and 
						$regtypes{${$generegion}{$gene}}
						)
					)){
				#print "extracting\n";
				
				# If unrecognised
				if( not ${$genesort}{$gene}){
					push @{$unrec_genes{$gene}}, $seqname;
					$checked_features++ and next unless $writeunknowns;
				}
				
				# Make a new output object for this gene if needed
				if(! $faobjs{$gene}){
					$faobjs{$gene} = Bio::SeqIO->new(-file => ">${outdir}/$gene.fa",
									 -format => "fasta");
				}
				
				# Check if gene sequence is truncated
				my $codon_start = 1;
				if($feat->has_tag('codon_start') and $keepframe){
					my @values = $feat->get_tag_values("codon_start");
					die "Error: $seqname $feattype annotation $gene has multiple \\codon_start values!\n" if(scalar @values > 1);
					$codon_start = $values[0]
				}
				
				# Compile output sequence object and store
				my $outseq = $feat->seq;
				$outseq = $outseq->trunc($codon_start, $outseq->length());
				if($organism){
					$outseq->display_id($species) or $outseq->display_id($seqname);
				} else {
					$outseq->display_id($seqname);
				}
				$outseq->description("");
				$outseq->verbose(-1);
				
				$found_genes{$gene} = $outseq;
				
			} else {
				#print "skipping\n";
			}
			$checked_features++;
		}
		
		# Run checks against threshold number and content of found genes
		my @found_gene_names = keys %found_genes;
		my $n_found_known_genes = scalar (intersect( @found_gene_names, @{$known_genes}));
		warn "Sequence $seqname in $gbp has $n_found_known_genes known features with sufficient information, it will be skipped\n" and next unless $n_found_known_genes >= $minregion;
		
		warn "Sequence $seqname in $gbp does not include all required regions, it will be skipped\n" and next if scalar (intersect(@reqregion, @found_gene_names)) < scalar @reqregion and @reqregion;
		
		# Write out found genes
		$faobjs{$_}->write_seq($found_genes{$_}) foreach(keys %found_genes);
		
	}
}

if(scalar keys %unrec_genes > 0){
	print "Unrecognised gene names:\n", join (", ", keys %unrec_genes), "\n\nSource sequences of unrecognised gene names:\n" ;
	
	foreach my $ugene (keys %unrec_genes){
		print "$ugene: ", join (", ", @{$unrec_genes{$ugene}}), "\n";
	}
}

exit;

sub sort_genes{
	my ($genenametxt) = @_;
	my @genenames;
	my %genesort;
	my %generegion;
	foreach my $line (split("\n", $genenametxt)){
		my ($name, $region, $product, @values) = split("[;:,]", $line);
		push @values, ($name, $product);
		print Dumper \@values;
		push @genenames, $name;
		$generegion{$name} = $product;
		foreach my $var (@values){
			$genesort{$var} = $name;
		}
	}
	return(\@genenames, \%genesort, \%generegion)
}

sub get_clean_name{
	my ($feat) = @_;
	my $name;
	
	if($feat->has_tag('gene')){
		($name) = $feat->get_tag_values('gene');
	} elsif($feat->has_tag('label')){
		($name) = $feat->get_tag_values('label');
	}
	if($name){
		$name = uc $name;
		$name =~ s/[(;_ ].*$//;
		$name =~ s/(?<!MT)-.*$//;
	} elsif($feat->has_tag('product')){
		($name) = $feat->get_tag_values('product');
		$name = uc $name;
	} else {
		die "Error: feature passed to get_clean_name subroutine without gene, label or product tag\n";
	}
	$name =~ s/[\/,\\:;]/_/g;
	return($name)
}

File: extract_genes.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Description of what the script does"""

# Imports

import argparse
import textwrap as _textwrap
import urllib
import os
import sys
from collections import defaultdict

from Bio import SeqIO


# Class definitions

# This reclasses the argparse.HelpFormatter object to have newlines in the help text for paragrahps
class MultilineFormatter(argparse.HelpFormatter):
    def _fill_text(self, text, width, indent):
        text = self._whitespace_matcher.sub(' ', text).strip()
        paragraphs = text.split('|n ')
        multiline_text = ''
        for paragraph in paragraphs:
            formatted_paragraph = _textwrap.fill(paragraph, width, initial_indent=indent,
                                                 subsequent_indent=indent
                                                 ) + '\n\n'
            multiline_text = multiline_text + formatted_paragraph
        return multiline_text


# Global variables

# Function definitions

def loadnamevariants(report=False):
    if report:
        print("The following are the gene name variants currently known.\n")
    output = {}
    fullparse = {}
    alltypes = set()
    url = "https://raw.githubusercontent.com/tjcreedy/constants/master/gene_name_variants.txt"
    for line in urllib.request.urlopen(url):
        line = line.decode('utf-8').strip()
        description, variants = line.split(":")
        name, annotype, fullname = description.split(";")
        variants = variants.split(',')
        variants.extend([name, fullname.upper()])
        fullvariants = []
        for v in variants:
            for g in ['', ' ']:
                v = v.replace(g, '')
                for s in ['', ' GENE', ' '+annotype.upper()]:
                    var = v+s
                    fullvariants.append(var)
                    output[v+s] = name
        alltypes.add(annotype)
        fullparse[name] = {'type': annotype, 'variants': fullvariants}
        if report:
            fullvariants = [v.replace(' ', '\u00A0') if len(v) < 12 else v for v in fullvariants]
            fullvariants = _textwrap.fill(', '.join(fullvariants), width=80,
                                          initial_indent='\t', subsequent_indent='\t')
            print(f"Standard name = {name}, type = {annotype}, full name = {fullname}:\n"
                  f"{fullvariants}")
    if not report:
        return output, alltypes, fullparse


def get_feat_name(feat):
    featname = "unknown"
    nametags = ['gene', 'product', 'label', 'standard_name']
    if any(t in feat.qualifiers.keys() for t in nametags):
        for t in nametags:
            if t in feat.qualifiers.keys():
                featname = feat.qualifiers[t][0].upper()
                break
    return featname


def getcliargs(arglist=None, knowngenes=None, knowntypes=None):
    # Initialise the parser object and give a description
    parser = argparse.ArgumentParser(description="""
    This script finds and extracts sequences corresponding to CDS, rRNA and/or tRNA annotations 
    from a set of sequences from one or more genbank-format flat files. Specify which regions to 
    extract using the -r/--regiontypes argument (default is CDS)
    |n
    The region name is identified based on the /gene, /label or /product tag in the genbank-format 
    flat file. Two methods are used to ensure that naming variants are correctly identified as the 
    same gene:
        1. the gene name is converted to uppercase (so that "atp8" is the same as "ATP8")
        2. the script removes any semicolons, underscores, hyphens or spaces, and any characters 
        following these, from gene names ("ATP8-0" -> "ATP8", "ATP8; atp8" -> "ATP8", etc) - 
        except if the first part of the name is "MT-".
    |n
    Additionally, a library of common naming variants of mitochondrial regions are retrived from 
    https://github.com/tjcreedy/constants. For example, CDS regions COX1 and COI will both be 
    output to COX1. The variants retrieved can be reviewed using the -s/--showgenes option.
    |n
    A separate fasta will be written for each region found, containing all of the sequences of 
    that CDS from across all the different sequences in the genbank file(s). The LOCUS name of the 
    sequence in the genbank-format flat file will be used as the sequence header. Alternatively, 
    you can try to use use the organism field of the genbank file as the sequence header using 
    -o/--organism: note that this is very likely to give blank or non-unique headers, and if the 
    script fails to parse the species name it will revert to the LOCUS field.
    |n
    Optionally, you can specify that sequences will only be output for entries where a minimum 
    number of different genes are met using -m/--mingene. Similarly, you can optionally specify that 
    certain genes must be present in a sequence for it to output any genes using the 
    -q/--reqregion option. For example, if you only wanted to retrieve sequences for regions from 
    genomes that contain COX1 and COB, you would specify --reqregion COX1 COB. The names used must 
    conform with the standard set of output names (the second column when using the 
    --showgenes option). All available regions will still be output.
    |n
    Also optionally, you can output a list of genes present for each sequence by supplying the 
    path to the output text file name to -p/--presence.
    |n
    In rare cases, an annotation is truncated by the end of a contig at the 5' end, which may 
    leave an incomplete codon at the start of an output sequence and cause downstream issues. To 
    detect this, the script can optionally remove these excess 1 or 2 bases using the 
    -k/---keepframe option. Note this only works if the /codon_start tag is present in the 
    genbank-format file features. 
    """, formatter_class=MultilineFormatter)

    # Add individual argument specifications
    parser.add_argument("-g", "--genbank", type=str, metavar='PATH', required=True, nargs='+',
                        help="the path(s) to one or more genbank files from which gene sequences "
                             "should be extracted")
    parser.add_argument("-o", "--output", type=str, metavar='PATH', required=True,
                        help="the path to a directory in which to place a fasta file for each "
                             "gene, this will be created if it doesn't exist")
    parser.add_argument("-m", "--mingenes", type=int, metavar='N', default=0,
                        help="the minimum number of genes that must be present in an entry for "
                             "any genes to be output")
    parser.add_argument("-q", "--reqgenes", type=str, metavar='GENE', nargs='*',
                        help="a list of genes that must be present in an entry for any gene "
                             "sequences to be output",
                        choices=knowngenes)
    parser.add_argument("-r", "--genetypes", type=str, metavar='TYPE', nargs='+',
                        help="one or more gene types to be output",
                        choices=knowntypes, default=knowntypes)
    parser.add_argument("-f", "--filter", type=str, metavar='PATH',
                        help="path to a file listing LOCUS names, one per line; if supplied, only "
                             "these entries will be output")
    parser.add_argument("-n", "--organism", action='store_true',
                        help="use the organism field instead of the locus field for sequence "
                             "headers in the output files")
    parser.add_argument("-w", "--writeunknowns", action='store_true',
                        help="write fastas for unidentifiable annotations anyway")
    parser.add_argument("-k", "--keepframe", action='store_true',
                        help="remove excess out-of-frame bases at the beginning of truncated "
                             "annotations")
    parser.add_argument("-p", "--presence", type=str, metavar='PATH',
                        help="if desired, the path to a text file to which will be written a list "
                             "of genes present for each input entry")
    parser.add_argument("-s", "--showgenes", action='store_true',
                        help="print the known gene name variants")

    # Parse the arguments from the function call if specified, otherwise from the command line
    args = parser.parse_args(arglist) if arglist else parser.parse_args()

    # Do some checking of the inputs
    if args.mingenes < 0:
        parser.error(f"{args.mingenes} is not a possible minimum number of genes")

    # If the arguments are all OK, output them
    return args


# Start the actual script
if __name__ == "__main__":

    # Get the gene name variants
    nameconvert, annotypes, namevariants = loadnamevariants()

    # Get the arguments
    args = getcliargs(None, nameconvert.keys(), annotypes)
    #args = getcliargs('-g /home/thomas/work/iBioGen_postdoc/MMGdatabase/newdata_2022-04-15_syrphid/testnewsequences.gb -o testextract -w -k -p testpresence.txt'.split(' '),  nameconvert.keys(), annotypes)  # Read from a string, good for testing

    # Print gene name variants
    if args.showgenes:
        loadnamevariants(report=True)
        exit()

    # Make output directory if necessary
    if not os.path.exists(args.output):
        os.makedirs(args.output)

    # Set up for unrecognised genes
    unrecgenes = defaultdict(list)

    # Set up gene presence handle
    if args.presence:
        args.presence = open(args.presence, 'w')

    # Read filter list
    filter = []
    if args.filter:
        with open(args.filter) as fh:
            for line in fh:
                filter.append(line.rstrip())

    # Work through genbank files
    nrejected = 0
    outfh = {}
    for gbpath in args.genbank:
        # gbpath = args.genbank[0]
        for seqrecord in SeqIO.parse(gbpath, "genbank"):
            # gb = SeqIO.parse(gbpath, "genbank")
            # seqrecord = next(gb)
            # If filtering, check if this entry should be used
            if args.filter and seqrecord.name not in filter:
                nrejected += 1
                continue

            # Extract names
            seqname = seqrecord.name
            outname = seqname
            if args.organism and seqrecord.organism:
                outname = seqrecord.organism.replace(' ', '_')

            # Set up containers
            foundgenes = defaultdict(list)

            # Work through features
            for feat in seqrecord.features:
                # feat = seqrecord.features[3]
                # Skip if not requested type
                if feat.type not in args.genetypes:
                    continue

                # Convert feat name
                name = get_feat_name(feat)
                if name in nameconvert:
                    stdname = nameconvert[name]
                else:
                    unrecgenes[name].append(seqname)
                    if args.writeunknowns:
                        stdname = name
                    else:
                        continue

                # Store sequence for writing
                featsequence = feat.extract(seqrecord.seq)
                if args.keepframe and 'codon_start' in feat.qualifiers:
                    featsequence = featsequence[(int(feat.qualifiers['codon_start'][0])-1):]
                foundgenes[stdname].append(featsequence)

            # Check to see if there are enough found features and all required genes are present,
            # if so write sequences out
            if len(foundgenes) >= args.mingenes:
                if not args.reqgenes or all([g in foundgenes for g in args.reqgenes]):
                    for gene, seqs in foundgenes.items():
                        # Check the number of sequences and warn if there are > 1
                        if len(seqs) > 1:
                            sys.stderr.write(f"Warning: entry {seqname} has multiple annotations "
                                             f"of the target type ({'/'.join(args.genetypes)}) "
                                             f"matching the standard name \"{gene}\"\n")
                        # Check to see if there's already a file handle for this gene, if not open
                        if gene not in outfh:
                            outfh[gene] = open(os.path.join(args.output, f"{gene}.fasta"), 'w')
                        # Write out sequences
                        for seq in seqs:
                            outfh[gene].write(f">{outname}\n{seq}\n")

                    if args.presence:
                        args.presence.write(f"{outname},{','.join(foundgenes.keys())}\n")

                else:
                    sys.stderr.write(f"Warning: entry {seqname} does not include all required "
                                     f"genes, it will be skipped")
            else:
                sys.stderr.write(f"Warning: entry {seqname} has {len(foundgenes)} known "
                                 f"annotations with sufficient information, it will be skipped")

    # Close file handles
    for gene, fh in outfh.items():
        fh.close()
    if args.presence:
        args.presence.close()

    # Report unrecognised genes
    if len(unrecgenes) > 0:
        sys.stderr.write(f"Warning: unrecognised genes present in the following entries.\n")
        for gene, entries in unrecgenes.items():
            sys.stderr.write(f"\t{gene} - {', '.join(sorted(list(set(entries))))}\n")

    # Report rejected entries
    if nrejected > 0:
        sys.stderr.write(f"Warning: {nrejected} entries in genbank-format file(s) rejected for "
                         f"not matching filter list")
